Mailgun integration

zemccartney opened the pull request:
@nelsonic heyo! got sending with Mailgun working, tests passing, all linted

I still need to cleanup the README, specifically:

- the Background Reading section for Mailgun is plain inaccurate :) sorry! This in fact uses Mailgun's official library, not the 3rd party one
- I want to add a section describing how to use AWS and Mailgun (or any other non-SES mailing service, if added in the future) in the same deployment by specifying the CURRENT_SERVICE env var (basically, the way I set lib/service.js up, it tries to guess the service someone wants to use based on their env vars; in the case of multiple validly configured services, it goes with the last service alphabetically (SES); if, for example, someone was using S3, so would have AWS' standard env vars set, but Mailgun for emailing, the tool would try (and fail) to use S3 instead; CURRENT_SERVICE allows specifying the service you want to use, overriding that default behavior

but I figure I'll hold off on that final documentation cleanup until after handling whatever revisions you want here, once the code's in a place you're happy with

thanks again for letting me take this on! 🙏

zemccartney commented:
Ay! My bad on the Travis checks. I probably should've thought about needing to update the .env file there :) 

Do I need to handle updating that or do you prefer to? Whichever's good! let me know if I can help with that

zemccartney commented:
@nelsonic heyo! do you need anything more from me here? or all set to review?

No rush at all! I know you're busy and this is a side-thing, checking in mostly so I don't forget about this. If the answer's just that I should be patient, I can do that, too :)

Hope you had a great weekend!

zemccartney commented:
Heyo @nelsonic ! anything I can do to push this forward?

again, i know it's small potatoes, no big rush, mainly curious :) Thanks again for the help here. Hope you've been having a good weekend 🙏

zemccartney commented:
@samhstn hey! sorry, I'm super late here, but I finally cleaned up that merge conflict, all tests are back to running, and README is up-to-date.

Travis checks aren't passing, I assume, because some new env vars need to be set for Mailgun. Lemme know if I can help with updating that at all (i.e. if my code / instructions ended up totally unclear).

How is this looking to you?

samhstn commented on lib/index.js:
```diff
@@ -68,58 +65,30 @@ function isTruthy (x) {
 
-  ses.sendEmail(params, callback);
+  // Deep clones options, ensures no mutation of input (deep necessary because source may contain arrays)
+  var clean_options = JSON.parse(JSON.stringify(options));
```
@zemccartney could you just explain why a deep clone was needed here? As far as I'm aware we weren't mutating the options passed in before. At which point did this become necessary? - I just think it would be great if it was possible to avoid this.

(I quite liked how easy these lines were to read before: https://github.com/dwyl/sendemail/blob/master/lib/index.js#L77-L85)

zemccartney replied:
@samhstn hey, thanks calling this out. definitely a bit of a kludge. my bad on that.

I used a deep clone here because I needed to end up with an object containing all the values of the `options` object that I could then pass as the first parameter to the `send` call. Since we're running those conditional checks and filters on the properties of that object and need to store those results, I _think_ I thought that meant we'd be mutating the input `options` (sorry, it's been long enough since I wrote this that I've since forgotten / am a bit lost myself on what I was thinking :) ). 

I tried to clean this up a good bit in the work I just pushed up. Still creating a new object, but not cloning our options object. Hopefully readable, little less misdirection here. How does that look?

samhstn commented on lib/service.js:
```diff
@@ -0,0 +1,166 @@
+ * @returns {Object} interface for parameterizing and sending email send requests
+ * with the determined service
+ */
+function determine_service (specific_service) {
```
I have a feeling there is maybe a nicer way to do this, having this as part of the library code and referencing another `CURRENT_SERVICE` env var feels a little off, let me know your thoughts @nelsonic @zemccartney?

zemccartney replied:
@samhstn totally agree, I was never all that happy with how I did things here. 

I came up with a new approach to this in the stuff I just pushed up. 

I still have the library using another env var, now renamed to `SENDEMAIL_SERVICE` to hopefully make the var's name more descriptive/understandable, but only if the user happens to have multiple services configured at once.  In that case, we throw an error saying the user's env config is ambiguous, directing them to use that variable to tell the library which sending service to use. 

In short, I tried to clarify and minimize the purpose of that env var by moving it out of the main path of the `determine_service` path. 

I think I was solving for the, to be honest, nonsensical defaulting behavior I had in place to handle this case before  (essentially, unless the user had set `CURRENT_SERVICE`, the library would use the last available service, in alphabetical order....no good 😆)

Does that get to root of this issue? Or at least improve things a bit?

My bad if not! I just jumped to this approach to try working through another thought. 
I kept `determine_service` for now because I figure we need to handle the case — I hope a seriously unlikely one, but I guess still possible — that a user could configure multiple services at once, like our tests. If that's not important, I bet we can simplify this part of the library a good deal (though not sure of the clearest route for that aside from continuing to use an env var...would have to think about that)

zemccartney replied:
@samhstn yo! any thoughts here?

sorry, no rush at all. I don't mean to bug, more chiming back so wrapping this up doesn't completely fall off my radar (again :) ). Let me know if there's anything I can do to help move this forward. Thanks for all the help so far

samhstn replied:
I've read through the code, thanks for the improvements here with comments and the explanations - very helpful.

I just want to raise https://github.com/dwyl/sendemail/pull/83#pullrequestreview-189591046 as a concern before any further reviewing or work is done in this pr.

samhstn commented on test/service.test.js:
```diff
@@ -0,0 +1,69 @@
+// TODO Delete this test / Or rather, switch to using CURRENT_SERVICE
+// TODO Add a test for using CURRENT_SERVICE
+// TODO Ok assumption?????
+// Assumes .env used for testing has all available services configured simultaneously
```
@zemccartney what is your plan with these `TODO`s, are you planning to address these in this pr or should we open an issue to work on these in a later pr?

zemccartney replied:
heh...sorry for that sloppiness. those are all resolved, I did a pass over the service testing file. I added a handful of new stuff, but all things that make the tests more reliable and readable.

samhstn commented:
In regards to the environment vars, we currently just have these in travis:
```
AWS_ACCESS_KEY_ID 
AWS_REGION
AWS_SECRET_ACCESS_KEY
SENDER_EMAIL_ADDRESS
TEMPLATE_DIRECTORY
```

Looks like we just need to add these two for the tests to pass:
```
MAILGUN_API_KEY
MAILGUN_SENDING_DOMAIN
```

zemccartney commented:
Hey @samhstn ! Very cool, thanks for all those detailed comments. I appreciate you taking the time to review this work and write those up 🙏 

I'm going to go back and leave responses, but I just pushed up another pass that I _think_ addresses all the issues you called out. 

Basically,
- gets rid of that deep clone, made that options validating / massaging in `sendMany` more closely resemble how it was before, hopefully a bit cleaner.
- refactors how we handle multiple services registered simultaneously. I moved from essentially randomly picking one to throwing an error that directs the user to set an env var to declare which service they want to use. More on this
- cleans up the tests a bit, closes those open TODOs

FINALLY (sorry to drag on here :) ), to your comment about setting new env vars, that's correct, but you'll also need to add `MAILGUN_AUTHORIZED_RECIPIENT` ( see )
This is a bit kludgy, I admit, but is used in testing only. Because Mailgun doesn't have a success simulator equivalent that I know of and to send to any inbox while in sandbox mode you need to invite that inbox to be an authorized recipient from your Mailgun dashboard, I've been authorizing a personal email as a recipient, then setting the authorized recipient var to that inbox.  Does that work ok for you all?

shaneparsons commented:
Is this ready to be merged? I'd love to use this in a current project!

nelsonic commented:
@zemccartney code is looking good. 🎉 
@samhstn thanks for reviewing with a fine-tooth-comb. ❤️ 
@shaneparsons we need to ensure that we have the necessary environment variables defined on TravisCI to ensure that this build will _always_ pass on CI ... we need to add an "admin task" for this. ⏳

zemccartney commented:
@nelsonic awesome, glad to hear! thanks so much for the feedback 🙏 let me know if there's anything at all I can do to help with the CI task

zemccartney commented:
👍 awesome, thanks for pointing that out @samhstn sorry for my oversight there.

IIRC, way back when I first started this, I debated between using that library and this community-contributed one: https://github.com/bojand/mailgun-js

I think I switched to Mailgun's official one, assuming it'd be more reliable, but looks like even Mailgun's [official node tutorial](https://www.mailgun.com/blog/how-to-send-transactional-email-in-a-nodejs-app-using-the-mailgun-api) directs people to use the community library :) Doi. On reviewing that library, it looks like it's still maintained (at least relatively recent issues and commits).

So I'm thinking I'll switch out Mailgun's client for the community one, make whatever updates are necessary to get that working. As part of that work, I'll audit that project's dependencies, see if there are any known security issues there, contacting the maintainer about my findings as needed.

How does that sound to you?

samhstn commented:
Sounds good @zemccartney, yeah the community contributed npm module looks better to use 👍thanks for looking into this ❤️

zemccartney commented:
@samhstn heyo! sorry for the long delay on this, hit a busy stretch. The commit I just pushed up replaces the unmaintained mailgun client with the maintained one and updates dependencies based on the results of `npm audit` (just `tap-spec` needed bumping).

Additionally, I noticed how I'd written things previously was a bit inefficient — I had been determining a service and instantiating a sending client on every send — so I refactored the main export of the service module (the service-specific sending function) to enclose just a call to the determined service's sending method. The upside is that this is a bit more efficient, I imagine. Downside is that you can no longer switch your sending service mid-process without clearing `sendemail` from `require`'s cache and re-requiring it (which is what we do now in the main test file). Not strictly necessary at all! Happy to switch back to how things were, just something I felt was an improvement.

Any questions or anything I can do to help here, lemme know!

Thanks again for all the help so far 🙏

samhstn commented on test/index.test.js:
```diff
@@ -70,160 +73,267 @@ test(file + " compile .txt template", function (t) {
+  var service_test_description = file + " — " + (specific_service || "Default Sender") + ":";
+
+  test(service_test_description + " Force Fail in Email for a single email", function (t) {
+    email_test_startup(specific_service);
```
I see why you've done this `startup` and `teardown`, but it would be nice if we could make `email` a "purer" function which doesn't read env vars and simplify this test

zemccartney replied:
Hmmm...gotcha, yea, I'd have to think on that one. But thanks for the feedback, though. Agree that'd be a good move; I definitely made the tests a good bit more.... busy :)

zemccartney commented:
Wicked, thanks for the feedback @samhstn 🙏 I'll address those notes soon

zemccartney commented on README.md:
```diff
@@ -60,15 +63,15 @@ people with *specific needs*.
 + [ ] install the `sendemail` module from NPM
-+ [ ] Ensure that you have an AWS Account and have downloaded your credentials.
-+ [ ] set the required [*environment variable*](https://github.com/dwyl/learn-environment-variables) (*see below*)
++ [ ] Ensure that you have an an account on your email service of choice
```
Ay! Sorry, sloppy, fixed! Thanks for catching that.

zemccartney commented on README.md:
````diff
@@ -94,6 +99,16 @@ export AWS_ACCESS_KEY_ID=YOURKEY
+
+```sh
+export TEMPLATE_DIRECTORY=./examples/templates
+TODO Confirm Mailgun requires verifying email addresses
````
Hmm, weird, I think I'd addressed this in a previous commit; regardless, this is confirmed and now better documented. 🙏

zemccartney commented on README.md:
```diff
@@ -143,19 +158,19 @@ email('welcome', person, function(error, result){
 For *full code of working example* see: [/examples/templates/**send-welcome-email.js**](https://github.com/dwyl/sendemail/blob/master/examples/send-welcome-email.js)
 Note: you still need to set your *environment variables*
-for the email to be sent.
+for the email to be sent. The example will work regardless of which service you use; our public API is platform agnostic TODO Is this true?
```
Confirmed! And I just added in a comment to clarify that the recipient address set in this example needs to be an authorized recipient if the user's sending domain is still sandboxed

zemccartney commented on README.md:
```diff
@@ -183,15 +198,26 @@ ___
+- low limit on the number of emails you can send (usually daily allowance)
+- you can send to and from only verified (manually whitelisted) email addresses, not any random email address from a random person stumbling across your site as in production use
+
+You're going to want to move out of this environment for sending email on a public project (in most cases)
```
Ay, yea, sure thing. What did you mean here, exactly? I tacked on a period at the end of the last sentence here; my bad if I misunderstood, though! Happy to revise further if I did 🤡

zemccartney commented:
@samhstn heyo! thanks again for all the feedback. I think I covered everything in the changes I just pushed up, but as always, let me know whatever else more is needed 🙏

samhstn commented:
@zemccartney looks great ❤️ thanks for your work

@nelsonic the [build is failing](https://travis-ci.org/dwyl/sendemail/builds/495265355?utm_source=github_status&utm_medium=notification) because `mailgun-js` uses the `class` syntax which is not supported on [`node` v4](https://github.com/dwyl/sendemail/blob/master/.travis.yml#L2), (see: https://stackoverflow.com/questions/41960142/block-scoped-declarations-not-yet-supported-outside-strict-mode)

What do you want to do about this? Upgrade version or look for another solution?

nelsonic commented on README.md:
```diff
@@ -63,7 +63,7 @@ people with *specific needs*.
 ### Checklist (*everything you need to get started in 5 minutes*)
 + [ ] install the `sendemail` module from NPM
-+ [ ] Ensure that you have an an account on your email service of choice
++ [ ] Ensure that you have an account for your email service of choice
```
👍

nelsonic commented:
@zemccartney we _agree_. LTS is the _preferred_ support target.
Our _original_ reasoning for having explicit Node v4 support was AWS Lambda.
We used sendemail on v4 on Lambda. 
The _current_ version of Node.js on AWS Lambda is **8.10**
https://docs.aws.amazon.com/lambda/latest/dg/lambda-runtimes.html
![image](https://user-images.githubusercontent.com/194400/53428122-1ccb7a00-39e2-11e9-9a07-f77bb2de849a.png)

I think we can safely discontinue support for node.js v.4 and even v.6 on the basis that v8.10 is available.

zemccartney commented:
@samhstn @nelsonic heyo! any word on moving this forward? not being impatient, no rush from me, just I realized I'd spaced on this, figured worth doubling back to see if there were anything I could do to help wrap this up. Thanks!

samhstn commented:
@nelsonic for our tests to pass, it looks like we need to set these two environment variables in `.travis.yml`:

```
MAILGUN_API_KEY
MAILGUN_SENDING_DOMAIN
```

Do you want me to go ahead and try and add them (I think I have permissions issues)?

zemccartney commented:
@samhstn yo! thanks for jumping right on this.

re: the new env vars, you'll also need to add `MAILGUN_AUTHORIZED_RECIPIENT`
This is a bit kludgy, I admit, but is used in testing only. Because Mailgun doesn't have a success simulator equivalent that I know of and to send to any inbox while in sandbox mode you need to invite that inbox to be an authorized recipient from your Mailgun dashboard, I've been authorizing a personal email as a recipient, then setting the authorized recipient var to that inbox. Does that work ok for you all?

zemccartney commented:
@nelsonic @samhstn heyo! sorry, been a while. just following up on this. is there anything I can do to help move this forward? as always, not to pressure at all. More to keep myself from forgetting :) Thanks!

zemccartney commented:
@nelsonic @samhstn hey y'all, checking back in on this. anything I can do here to move this forward?

I hope y'all've been well!

zemccartney commented:
Awesome, glad to hear, thanks for reviewing!
Anything I can help with on the env var front?

oleg-koval commented on .travis.yml:
```diff
@@ -1,6 +1,6 @@
 language: node_js
 node_js:
-  - 4.3.2
+  - 8.1
```
Shouldn’t it be >12?

nelsonic reviewed (approved):
Thanks @zemccartney. 👌 
Apologies that this one fell off the radar. 🤷‍♂️
